src/EventStore.Core/Services/Transport/Grpc/Enumerators.ReadAllBackwards.cs

```diff
@@ -43,96 +42,84 @@ public class ReadAllBackwards : IAsyncEnumerator<ResolvedEvent> {
 				}
 
 				_bus = bus;
-				_nextPosition = position;
 				_maxCount = maxCount;
 				_resolveLinks = resolveLinks;
 				_user = user;
 				_requiresLeader = requiresLeader;
 				_deadline = deadline;
-				_disposedTokenSource = new CancellationTokenSource();
-				_buffer = new ConcurrentQueue<ResolvedEvent>();
-				_tokenRegistration = cancellationToken.Register(_disposedTokenSource.Dispose);
-			}
+				_cancellationToken = cancellationToken;
+				_semaphore = new SemaphoreSlim(1, 1);
+				_channel = Channel.CreateBounded<ResolvedEvent>(BoundedChannelOptions);
 
+				ReadPage(position);
+			}
 
 			public ValueTask DisposeAsync() {
-				_disposedTokenSource.Dispose();
-				_tokenRegistration.Dispose();
-				return default;
+				_channel.Writer.TryComplete();
+				return new ValueTask(Task.CompletedTask);
 			}
 
 			public async ValueTask<bool> MoveNextAsync() {
-				if (_readCount >= _maxCount || _disposedTokenSource.IsCancellationRequested) {
+				if (_readCount >= _maxCount) {
 					return false;
 				}
 
-				if (_buffer.TryDequeue(out var current)) {
-					_current = current;
+				try {
+					_current = await _channel.Reader.ReadAsync(_cancellationToken).ConfigureAwait(false);
 					_readCount++;
 					return true;
-				}
+				} catch (ChannelClosedException ex) {
+					if (ex.InnerException is RpcException) {
```

Comment:
@thefringeninja Would it be better if we check for stream not found/default(null) exception to mean success and return false, or throw otherwise?